UCBChangesToScriptPics.js

```diff
@@ -0,0 +1,486 @@
+// BlockMorph menu:
+
+BlockMorph.prototype.userMenu = function () {
+    var menu = new MenuMorph(this),
+        world = this.world(),
+        myself = this,
+        hasLine = false,
+        shiftClicked = world.currentKey === 16,
+        proc = this.activeProcess(),
+        top = this.topBlock(),
+        vNames = proc && proc.context && proc.context.outerContext ?
+                proc.context.outerContext.variables.names() : [],
+        alternatives,
+        field,
+        rcvr;
+
+    function addOption(label, toggle, test, onHint, offHint) {
+        menu.addItem(
+            [
+                test ? new SymbolMorph(
+                    'checkedBox',
+                    MorphicPreferences.menuFontSize * 0.75
+                ) : new SymbolMorph(
+                    'rectangle',
+                    MorphicPreferences.menuFontSize * 0.75
+                ),
+                localize(label)
+            ],
+            toggle,
+            test ? onHint : offHint
+        );
+    }
+
+    function renameVar() {
+        var blck = myself.fullCopy();
+        blck.addShadow();
+        new DialogBoxMorph(
+            myself,
+            myself.userSetSpec,
+            myself
+        ).prompt(
+            "Variable name",
+            myself.blockSpec,
+            world,
+            blck.doWithAlpha(1, () => blck.fullImage()), // pic
+            InputSlotMorph.prototype.getVarNamesDict.call(myself)
+        );
+    }
+
+    menu.addItem(
+        "help...",
+        'showHelp'
+    );
+
+    /* UCB Script Pic Edit:
+     *  - Incorporated new method to userMenu to integrate scriptPic's and XML code.
+     *  - Code credited to Snap! Forum user Dardoro.
+     */
+    if (this.isTemplate || !this.definition) {
+		menu.addLine();
+		menu.addItem("script pic and xml...", 
+			() => {
```

Comment:
this is really nice! I would be more comfortable if this could be factored out into its own method rather than inlining it all into the code that represents the menu. We might even wish to reuse it for other image exports, right? I mean, we could export a sprite *as* picture and include all its code (and even other costumes, hehe!)

Reply:
I will definitely look into making this its own method so we can take care of any image exports! Which Morph do you think it would make the most sense to make this a prototype of since the method would have to work for multiple?

Reply:
Is there any reason not to include the xml in _all_ script pics?  That would reduce the cognitive load on users, and afaik doesn't make the png unreadable in other software.

Reply:
Yes I agree and that's definitely a possibility. I had it like that at first, but I thought I'd submit this PR consistent with @DarDoro's base project, though. If there are no objections, I also think that completely integrating it will be the best for user experience.

Reply:
I'm not so sure about this. There are - and will be - many screenshots and pictures of Snap! scripts in all kinds of web sites and print materials that don't offer embedded sources. Therefore I like that @DarDoro visually indicates the presence of sources in the generated image (but I'd also like to reflect on whether it should be a technical term like <xml> or perhaps some other pictogram / icon, perhaps a lambda or a stylized block, and we'll probably also place it somewhere where it doesn't overlap with any of the blocks, i.e. we'd want to add transparent space ontop).

Reply:
The original user case for this (i.e. curriculum authorship) sort of requires that we don't put a special label on images. We want them to look normal, since the use case is for authors, not students, to use the resulting scripts. 

Whether or not it should be everything is an interesting question. I am of two minds: On the one hand, including most scripts doesn't seem to hurt, but on the other hand we are going around turning a punch of static images into something semi-executable, at least inside Snap<em>!</em>. That's probably an OK trade off, but it's kind of like the dangers of allowing anyone to embed VB inside Word/Excel docs.

Reply:
The analog to malicious VB is a script pic with XML that doesn't match the picture.  I guess we _should_ worry about that, but once we _allow_ the PNG/XML combination, we've opened that door.  _Our_ script pics aren't going to be malicious in that way, so it really doesn't matter from a security standpoint if we _always_ make hybrid pictures or only _sometimes_ make them.  Bad guys will presumably take a PNG and add their own XML by hand.

But for the use case we have in mind, it's not necessary that we allow the XML to be run.  We have to allow it to be loaded, translated to another language, and exported as another script pic.  But we could invent the category of non-runnable script if that's what it takes to thwart bad guys.

Reply:
Yeah, that's true -- i mean really what we should just have is a command line script that converts images and we could delay the option to load scripts in.